src/hotspot/share/opto/escape.cpp

```diff
@@ -460,51 +460,66 @@ bool ConnectionGraph::can_reduce_phi_check_inputs(PhiNode* ophi) const {
       } else {
         ptn->set_scalar_replaceable(false);
       }
-    }
+   } else if (ophi->in(i)->is_Phi()) {
+      if (candidates.member(ophi->in(i)) || ophi->_idx == ophi->in(i)->_idx || can_reduce_phi_check_inputs(ophi->in(i)->as_Phi(), candidates)) {
+	  found_sr_allocate = true;
+      }  
+   }								   
   }
-
   NOT_PRODUCT(if (TraceReduceAllocationMerges && !found_sr_allocate) tty->print_cr("Can NOT reduce Phi %d on invocation %d. No SR Allocate as input.", ophi->_idx, _invocation);)
   return found_sr_allocate;
 }
 
 // Check if we are able to untangle the merge. Right now we only reduce Phis
 // which are only used as debug information.
-bool ConnectionGraph::can_reduce_phi_check_users(PhiNode* ophi) const {
+bool ConnectionGraph::can_reduce_phi_check_users(PhiNode* ophi, int nestedDepth) const {
   for (DUIterator_Fast imax, i = ophi->fast_outs(imax); i < imax; i++) {
     Node* use = ophi->fast_out(i);
 
-    if (use->is_SafePoint()) {
+    if (use->is_SafePoint() && nestedDepth < 2) {
       if (use->is_Call() && use->as_Call()->has_non_debug_use(ophi)) {
         NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. Call has non_debug_use().", ophi->_idx, _invocation);)
         return false;
       }
-    } else if (use->is_AddP()) {
+    } else if (use->is_AddP() && nestedDepth <= 2) {
       Node* addp = use;
       for (DUIterator_Fast jmax, j = addp->fast_outs(jmax); j < jmax; j++) {
         Node* use_use = addp->fast_out(j);
-        if (!use_use->is_Load() || !use_use->as_Load()->can_split_through_phi_base(_igvn)) {
-          NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. AddP user isn't a [splittable] Load(): %s", ophi->_idx, _invocation, use_use->Name());)
-          return false;
+        if (!use_use->is_Load() || !use_use->as_Load()->can_split_through_phi_base(_igvn, (nestedDepth > 1))) {
+         NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. AddP user isn't a [splittable] Load(): %s", ophi->_idx, _invocation, use_use->Name());)
+         return false;
         }
       }
+    } else if (nestedDepth > 1) {
+      // Note: Do not change the position of this else as all below cases are not eligible for nested optimizations
+      return false;
+    } else if (use->is_Phi()) {
+      if (ophi->_idx == use->_idx) {
+	tty->print_cr("***Can reduce self loop nested phi***");
```

Comment:
Does this really work? In any case, you need to check the debug flags before printing this message.